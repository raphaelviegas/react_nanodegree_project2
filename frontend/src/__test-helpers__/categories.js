export const initialState = { categories: [] }

export const categories = [
    {
        name: 'react',
        path: 'react'
    },
    {
        name: 'redux',
        path: 'redux'
    },
    {
        name: 'udacity',
        path: 'udacity'
    }
]