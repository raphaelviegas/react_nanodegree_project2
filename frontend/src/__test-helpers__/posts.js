const uuidv4 = require('uuid/v4')

export const initialState = {
    posts: [],
    sortByOption: '-voteScore',
    invalidPost: false,
    loadingPosts: true,
}

export const posts = [
    {
        id: '8xf0y6ziyjabvozdd253nd',
        timestamp: 1467166872634,
        title: 'Udacity is the best place to learn React',
        body: 'Everyone says so after all.',
        author: 'thingtwo',
        category: 'react',
        voteScore: 10,
        deleted: false,
        commentCount: 2
    },
    {
        id: '6ni6ok3ym7mf1p33lnez',
        timestamp: 1468479767190,
        title: 'Learn Redux in 10 minutes!',
        body: 'Just kidding. It takes more than 10 minutes to learn technology.',
        author: 'thingone',
        category: 'redux',
        voteScore: -5,
        deleted: false,
        commentCount: 0
    }
]

export const removePost = '8xf0y6ziyjabvozdd253nd'

export const votePost = {
    id: '8xf0y6ziyjabvozdd253nd',
    voteScore: 10
}

let id = uuidv4()
export const addPostInfo = {
    author: "Test Author",
    body: "Test Body",
    category: "Test",
    title: "Test Title",
    timestamp: Date.now(),
    id
}

export const editPostInfo = {
    id: '8xf0y6ziyjabvozdd253nd',
    editContent: {
        title: 'Edited post title',
        body: 'Edited post body'
    }
}