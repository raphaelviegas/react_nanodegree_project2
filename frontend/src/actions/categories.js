import * as ServerAPI from '../helpers/ServerAPI'

export const LOAD_CATEGORIES = 'LOAD_CATEGORIES'

//Local state
export function loadCategories(categories) {
  return {
    type: LOAD_CATEGORIES,
    payload: categories
  }
}


//Fetch requests
export const fetchCategories = () => dispatch => (
  ServerAPI.getCategories()
    .then(categories => dispatch(loadCategories(categories)))
);

