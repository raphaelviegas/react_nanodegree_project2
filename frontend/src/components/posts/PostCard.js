import React, { Component } from 'react'
import { Link } from 'react-router-dom'
import PropTypes from 'prop-types'
import MaterialIcon from 'material-icons-react'
import VoteScore from '../shared/VoteScore'
import EditPostForm from '../posts/editPost/EditPostForm'

class PostCard extends Component {
    static propTypes = {
        post: PropTypes.object.isRequired,
        onRemovePost: PropTypes.func.isRequired
    }

    state = {
        showEdit: '',
    }

    showEdit = () => this.setState(() => ({ showEdit: 'card-reveal-active' }))
    hideEdit = () => this.setState(() => ({ showEdit: '' }))

    removePost(id) {
        this.props.onRemovePost(id)
    }
    render() {
        const { id, title, body, author, voteScore, commentCount, category } = this.props.post
        const { showEdit } = this.state
        return (
            <div className="col s12 m6">
                <div className="card grey lighten-4">
                    <div className="card-content">
                        <span className="card-title truncate">{title}</span>
                        <span className="card-author">{author}</span>
                        <p>{body}</p>
                    </div>
                    <div className="card-action valign-wrapper">
                        <VoteScore size={16} score={voteScore} type='post' id={id} />
                        <Link className="align-right" to={`/${category}/${id}`}>Show Post</Link>
                    </div>
                    <div className="card-comments">
                        <span className="card-comments-number">{commentCount}</span>
                    </div>
                    <div className="card-buttons">
                        <div className="card-button-edit" onClick={() => this.showEdit()}>
                            <MaterialIcon icon="edit" size={14} />
                        </div>
                        <div className="card-button-delete" onClick={() => this.removePost(id)}>
                            <MaterialIcon icon="delete" size={14} />
                        </div>
                    </div>
                    <div className={`card-reveal ${showEdit}`}>
                        <EditPostForm
                            id={id}
                            defaultTitle={title}
                            defaultBody={body}
                            onPostCardFormSubmited={this.hideEdit}
                        />
                        <div className="hide-edit" onClick={() => this.hideEdit()}>
                            <MaterialIcon icon="close" color="#c3c3c3" />
                        </div>
                    </div>
                </div>
            </div>
        )
    }
}

export default PostCard
