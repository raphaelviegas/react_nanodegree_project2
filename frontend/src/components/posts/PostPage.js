import React, { Component } from 'react'
import { connect } from 'react-redux'
import { withRouter, Redirect } from 'react-router-dom'
import { fetchPosts, fetchRemovePost } from '../../actions/posts'
import MaterialIcon from 'material-icons-react'
import Moment from 'react-moment';
import VoteScore from '../shared/VoteScore'
import Comments from '../comments/Comments'
import EditPostModal from '../posts/editPost/EditPostModal'

export class PostPage extends Component {
    state = {
        modalOpen: ''
    }

    openModal = () => this.setState(() => ({ modalOpen: 'modal-open' }))
    closeModal = () => this.setState(() => ({ modalOpen: '' }))

    componentDidMount() {
        this.props.fetchPosts()
    }

    componentDidUpdate(prevProps) {
        if (this.props.id !== prevProps.id) {
            this.props.fetchPosts()
        }
    }

    removePost = (id) => {
        this.props.fetchRemovePost(id)
        this.props.history.push('/')
    }

    render() {
        const { postInfo } = this.props
        let invalidPost = (postInfo.length > 0) ? false : true
        return (
            <div>
                {invalidPost
                    ? <Redirect to="/error" />
                    : <div className="post-page">
                        {postInfo.map((post) => (
                            <div key={post.id}>
                                <div className="post-buttons">
                                    <div className="post-button-edit" onClick={() => this.openModal()}>
                                        <MaterialIcon icon="edit" size={20} />
                                    </div>
                                    <div className="post-button-remove" onClick={() => this.removePost(post.id)}>
                                        <MaterialIcon icon="delete" size={20} />
                                    </div>
                                </div>
                                <div className="container">
                                    <h4 className="post-title">{post.title}</h4>
                                    <div className="post-infos">
                                        <span className="post-author">Author: {post.author}</span>
                                        <span className="post-category">Category: {post.category}</span>
                                        <span className="post-date">
                                            <MaterialIcon icon="calendar_today" size={16} />
                                            <Moment format="DD/MMM/YY">{post.timestamp}</Moment>
                                        </span>
                                    </div>
                                    <VoteScore size={18} score={post.voteScore} type='post' id={post.id} />
                                    <p className="post-body">{post.body}</p>
                                    <Comments parentId={post.id} />
                                </div>
                                <EditPostModal
                                    modalOpen={this.state.modalOpen}
                                    onCloseModal={this.closeModal}
                                    id={post.id}
                                    title={post.title}
                                    body={post.body}
                                />
                            </div>
                        ))}
                    </div>
                }
            </div>
        )
    }
}
const mapStateToProps = (state, ownProps) => {
    return {
        id: ownProps.match.params.postId,
        postInfo: state.postsReducer.posts.filter(post => post.id === ownProps.match.params.postId)
    }
}

const mapDispatchToProps = (dispatch) => {
    return {
        fetchPosts: () => dispatch(fetchPosts()),
        fetchRemovePost: (id) => dispatch(fetchRemovePost(id))
    }
}

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(PostPage));
