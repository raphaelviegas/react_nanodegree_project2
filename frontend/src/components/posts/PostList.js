import React, { Component } from 'react'
import { connect } from 'react-redux'
import { withRouter } from 'react-router-dom'
import PropTypes from 'prop-types'
import { fetchRemovePost } from '../../actions/posts'
import PostCard from './PostCard'
import sortBy from 'sort-by'

export class PostList extends Component {
  static propTypes = {
    posts: PropTypes.array.isRequired,
    sortByOption: PropTypes.string.isRequired
  }

  removePost = (id) => {
    this.props.fetchRemovePost(id)
  }

  render() {
    const { posts, sortByOption } = this.props
    //SortBy
    const sortedPosts = [...posts].sort(sortBy(sortByOption))
    return (
      <div>
        {posts.length === 0
          ? <div className="noPosts"><h4>No posts to show</h4></div>
          : <div className="row">
            {
              sortedPosts.map((post, index) => {
                return <PostCard post={post} onRemovePost={this.removePost} key={index} />
              })
            }
          </div>
        }
      </div>
    )
  }
}

const mapStateToProps = (state, ownProps) => {
  return {
    posts: ownProps.posts,
    sortByOption: state.postsReducer.sortByOption
  }
}

const mapDispatchToProps = (dispatch) => {
  return {
    fetchRemovePost: (id) => dispatch(fetchRemovePost(id))
  }
}

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(PostList));
