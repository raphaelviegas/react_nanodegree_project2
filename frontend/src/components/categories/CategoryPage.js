import React, { Component } from 'react'
import { connect } from 'react-redux'
import PropTypes from 'prop-types'
import { fetchPosts } from '../../actions/posts'
import PostList from '../posts/PostList'
import OrderPosts from '../shared/OrderPosts'

export class CategoryPage extends Component {
    static propTypes = {
        category: PropTypes.string.isRequired,
        categories: PropTypes.array.isRequired,
        posts: PropTypes.array.isRequired
    }

    verifyCategory() {
        if (this.props.categories.length > 0) {
            let categoryExists = false
            // eslint-disable-next-line
            this.props.categories.map(item => {
                if (item.path === this.props.category) {
                    categoryExists = true
                }
            })
            if (categoryExists === false) {
                this.props.history.push('/error')
            }
        }
    }

    componentDidMount() {
        this.props.fetchPosts()
        this.verifyCategory()
    }

    componentDidUpdate(prevProps) {
        if (this.props.category !== prevProps.category) {
            this.props.fetchPosts()
            this.verifyCategory()
        }
    }

    render() {
        const { posts, category } = this.props
        const categoryPosts = posts.filter(post => post.category === category)
        return (
            <div>
                <div className="category-title">
                    Category: <span>{this.props.category}</span>
                    <OrderPosts />
                </div>
                <div className="divider"></div>
                <PostList posts={categoryPosts} />
            </div>
        )
    }
}

const mapStateToProps = (state, ownProps) => {
    return {
        category: ownProps.match.params.category,
        categories: state.categoriesReducer.categories,
        posts: state.postsReducer.posts,
    }
}

const mapDispatchToProps = (dispatch) => {
    return {
        fetchPosts: () => dispatch(fetchPosts())
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(CategoryPage)
