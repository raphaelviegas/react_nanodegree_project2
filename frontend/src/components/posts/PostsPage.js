import React, { Component } from 'react'
import { connect } from 'react-redux'
import PropTypes from 'prop-types'
import { fetchPosts } from '../../actions/posts'
import PostList from '../posts/PostList'
import OrderPosts from '../shared/OrderPosts'

export class PostsPage extends Component {
    static propTypes = {
        posts: PropTypes.array.isRequired
    }

    componentDidMount() {
        this.props.fetchPosts()
        let elems = document.querySelectorAll('.dropdown-trigger');
        // eslint-disable-next-line
        let instances = M.Dropdown.init(elems, { alignment: 'right', constrainWidth: false })
    }

    render() {
        return (
            <div>
                <div className="category-title">
                    Showing All Posts
                <OrderPosts />
                </div>
                <div className="divider"></div>
                <PostList posts={this.props.posts} />
            </div>
        )
    }
}

const mapStateToProps = (state) => {
    return {
        posts: state.postsReducer.posts,
    }
}

const mapDispatchToProps = (dispatch) => {
    return {
        fetchPosts: () => dispatch(fetchPosts())
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(PostsPage)
