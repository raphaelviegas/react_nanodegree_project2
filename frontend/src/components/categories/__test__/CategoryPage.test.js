import React from "react"
import { shallow, mount, render } from "enzyme"
import { CategoryPage } from '../CategoryPage'
import { categories } from '../../../__test-helpers__/categories'
import { posts } from '../../../__test-helpers__/posts'


describe("<CategoryPage />", () => {
    // mock function to replace the one provided by mapDispatchToProps
    const mockfetchPosts = jest.fn()
    const mockverifyCategories = jest.fn()
    const testCategory = 'react'
    const setup = {
        posts,
        categories,
        category: testCategory,
        fetchPosts: mockfetchPosts
    }
    let wrapper
    beforeEach(() => {
        mockfetchPosts.mockClear()
        mockverifyCategories.mockClear()
        wrapper = shallow(<CategoryPage {...setup} />)
        wrapper.instance().verifyCategory = mockverifyCategories
    })

    it('Should render correctly', () => {
        expect(wrapper).toMatchSnapshot()
    })

    it('Should call fetchPosts and verifyCategory when component mounts', () => {
        expect(mockfetchPosts).toHaveBeenCalledTimes(1)
    })

    it('Should call fetchPosts when component updates', () => {
        wrapper.setProps({ category: 'redux' })
        expect(mockfetchPosts).toHaveBeenCalledTimes(2)
    })

    it('Should call verifyCategory() when component mounts', () => {
        wrapper.instance().componentDidMount()
        expect(mockverifyCategories).toHaveBeenCalledTimes(1)
    })

    it('Should call verifyCategory() when component updates', () => {
        wrapper.setProps({ category: 'redux' })
        expect(mockverifyCategories).toHaveBeenCalledTimes(1)
    })

    it('Should not call verifyCategory() when component updates if the props are the same', () => {
        wrapper.setProps({ category: testCategory })
        expect(mockverifyCategories).toHaveBeenCalledTimes(0)
    })

    it('Should render a Connected OrderPosts component', () => {
        const componentOrderPosts = wrapper.find('Connect(OrderPosts)')
        expect(componentOrderPosts).toHaveLength(1)
    })

    it('Should find a withRouter(Connect(PostList)) with posts props', () => {
        let componentPostList = wrapper.find('withRouter(Connect(PostList))')
        expect(componentPostList).toHaveLength(1)
        const filteredPosts = posts.filter(post => post.category === testCategory)
        expect(componentPostList.prop('posts')).toEqual(filteredPosts)
    })

    it('Should redirect user to error page when category doesnt exists', () => {
        const historyMock = { push: jest.fn() };
        const newSetup = {
            posts,
            categories,
            category: 'invalidCategory',
            fetchPosts: mockfetchPosts,
        }
        const newWrapper = shallow(<CategoryPage {...newSetup} history={historyMock} />)
        expect(historyMock.push).toHaveBeenCalledTimes(1)
    })
})
