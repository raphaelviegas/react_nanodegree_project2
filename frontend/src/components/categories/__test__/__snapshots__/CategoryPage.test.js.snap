// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CategoryPage /> Should render correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <CategoryPage
    categories={
        Array [
            Object {
              "name": "react",
              "path": "react",
            },
            Object {
              "name": "redux",
              "path": "redux",
            },
            Object {
              "name": "udacity",
              "path": "udacity",
            },
          ]
    }
    category="react"
    fetchPosts={[Function]}
    posts={
        Array [
            Object {
              "author": "thingtwo",
              "body": "Everyone says so after all.",
              "category": "react",
              "commentCount": 2,
              "deleted": false,
              "id": "8xf0y6ziyjabvozdd253nd",
              "timestamp": 1467166872634,
              "title": "Udacity is the best place to learn React",
              "voteScore": 10,
            },
            Object {
              "author": "thingone",
              "body": "Just kidding. It takes more than 10 minutes to learn technology.",
              "category": "redux",
              "commentCount": 0,
              "deleted": false,
              "id": "6ni6ok3ym7mf1p33lnez",
              "timestamp": 1468479767190,
              "title": "Learn Redux in 10 minutes!",
              "voteScore": -5,
            },
          ]
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="category-title"
>
          Category: 
          <span>
                    react
          </span>
          <Connect(OrderPosts) />
</div>,
        <div
          className="divider"
/>,
        <withRouter(Connect(PostList))
          posts={
                    Array [
                              Object {
                                "author": "thingtwo",
                                "body": "Everyone says so after all.",
                                "category": "react",
                                "commentCount": 2,
                                "deleted": false,
                                "id": "8xf0y6ziyjabvozdd253nd",
                                "timestamp": 1467166872634,
                                "title": "Udacity is the best place to learn React",
                                "voteScore": 10,
                              },
                            ]
          }
/>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "Category: ",
            <span>
              react
</span>,
            <Connect(OrderPosts) />,
          ],
          "className": "category-title",
        },
        "ref": null,
        "rendered": Array [
          "Category: ",
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "react",
            },
            "ref": null,
            "rendered": "react",
            "type": "span",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {},
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "divider",
        },
        "ref": null,
        "rendered": null,
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "posts": Array [
            Object {
              "author": "thingtwo",
              "body": "Everyone says so after all.",
              "category": "react",
              "commentCount": 2,
              "deleted": false,
              "id": "8xf0y6ziyjabvozdd253nd",
              "timestamp": 1467166872634,
              "title": "Udacity is the best place to learn React",
              "voteScore": 10,
            },
          ],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="category-title"
>
            Category: 
            <span>
                        react
            </span>
            <Connect(OrderPosts) />
</div>,
          <div
            className="divider"
/>,
          <withRouter(Connect(PostList))
            posts={
                        Array [
                                    Object {
                                      "author": "thingtwo",
                                      "body": "Everyone says so after all.",
                                      "category": "react",
                                      "commentCount": 2,
                                      "deleted": false,
                                      "id": "8xf0y6ziyjabvozdd253nd",
                                      "timestamp": 1467166872634,
                                      "title": "Udacity is the best place to learn React",
                                      "voteScore": 10,
                                    },
                                  ]
            }
/>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "Category: ",
              <span>
                react
</span>,
              <Connect(OrderPosts) />,
            ],
            "className": "category-title",
          },
          "ref": null,
          "rendered": Array [
            "Category: ",
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "react",
              },
              "ref": null,
              "rendered": "react",
              "type": "span",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {},
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "divider",
          },
          "ref": null,
          "rendered": null,
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "posts": Array [
              Object {
                "author": "thingtwo",
                "body": "Everyone says so after all.",
                "category": "react",
                "commentCount": 2,
                "deleted": false,
                "id": "8xf0y6ziyjabvozdd253nd",
                "timestamp": 1467166872634,
                "title": "Udacity is the best place to learn React",
                "voteScore": 10,
              },
            ],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
